test/cpp/qps/client_sync.cc

```diff
@@ -182,12 +181,40 @@ class SynchronousStreamingClient : public SynchronousClient {
     // don't set the value since the stream is failed and shouldn't be timed
     entry->set_status(s.error_code());
     if (!s.ok()) {
-      gpr_log(GPR_ERROR, "Stream %" PRIuPTR " received an error %s", thread_idx,
-              s.error_message().c_str());
+      std::lock_guard<std::mutex> l(stream_mu_[thread_idx]);
+      if (!shutdown_[thread_idx].val) {
+        gpr_log(GPR_ERROR, "Stream %" PRIuPTR " received an error %s",
+                thread_idx, s.error_message().c_str());
+      }
     }
+    // Lock the stream_mu_ now because the client context could change
+    std::lock_guard<std::mutex> l(stream_mu_[thread_idx]);
     context_[thread_idx].~ClientContext();
     new (&context_[thread_idx]) ClientContext();
   }
+  void OnAllStreams(std::function<bool(ClientContext*, StreamType*)> cleaner) {
+    std::vector<std::thread> cleanup_threads;
+    for (size_t i = 0; i < num_threads_; i++) {
+      cleanup_threads.emplace_back([this, i, cleaner]() {
+        std::lock_guard<std::mutex> l(stream_mu_[i]);
+        if (stream_[i]) {
+          shutdown_[i].val = cleaner(&context_[i], stream_[i].get());
```

Comment:
I'm a little confused by this.  What happens if stream is not initialized by the time we call ```OnAllStreams```?  Then shutdown would never get set.  It looks like this PR addresses a race condition where we shutdown before we finish initialization, but if looks like its impossible for that to happen.

Reply:
Good point, I guess that case wasn't happening yet in practice. I've resolved this.